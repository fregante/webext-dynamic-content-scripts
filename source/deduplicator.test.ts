import {
	test, describe, afterEach, expect,
} from 'vitest';
import {excludeDuplicateFiles} from './deduplicator.js';

const warnMock = jest.spyOn(console, 'warn').mockImplementation(() => undefined);

afterEach(() => {
	warnMock.mockClear();
});

describe('excludeDuplicateFiles', () => {
	test('safe', () => {
		expect(excludeDuplicateFiles([
			{
				js: ['first.js'],
				matches: ['https://virgilio.it/*'],
			},
			{
				js: ['first.js', 'second.js'],
				matches: ['https://example.com/*'],
			},
		])).toMatchSnapshot('it should remove duplicate scripts');
		expect(warnMock).not.toHaveBeenCalled();

		expect(excludeDuplicateFiles([
			{
				css: ['first.css'],
				matches: ['https://virgilio.it/*'],
			},
			{
				css: ['first.css', 'second.css'],
				matches: ['https://example.com/*'],
			},
		])).toMatchSnapshot('it should remove duplicate stylesheets');
		expect(warnMock).not.toHaveBeenCalled();

		expect(excludeDuplicateFiles([
			{
				js: ['alpha.js'],
				css: ['first.css'],
				matches: ['https://virgilio.it/*'],
			},
			{
				css: ['first.css', 'second.css'],
				js: ['alpha.js'],
				matches: ['https://example.com/*'],
			},
		])).toMatchSnapshot('it should remove duplicate scripts and stylesheets');
		expect(warnMock).not.toHaveBeenCalled();

		expect(excludeDuplicateFiles([
			{
				js: ['alpha.js'],
				css: ['first.css'],
				matches: ['https://virgilio.it/*'],
			},
			{
				css: ['first.css'],
				js: ['alpha.js'],
				matches: ['https://example.com/*'],
			},
		])).toMatchSnapshot('it should drop the whole block if if empty');
		expect(warnMock).not.toHaveBeenCalled();
	});
	test('warning', () => {
		expect(excludeDuplicateFiles([
			{
				js: ['first.js'],
				matches: ['https://virgilio.it/*'],
				exclude_matches: ['https://*/admin/*'],
			},
			{
				js: ['first.js'],
				exclude_matches: ['https://*/admin/*'],
				matches: ['https://*.example.com/*'],
			},
		])).toMatchSnapshot('it should not warn when a differentiator is the same');
		expect(warnMock).not.toHaveBeenCalled();

		expect(excludeDuplicateFiles([
			{
				css: ['first.css'],
				matches: ['https://virgilio.it/*'],
			},
			{
				css: ['first.css', 'second.css'],
				run_at: 'document_start',
				matches: ['https://example.com/*'],
			},
		])).toMatchSnapshot('it should warn when a differentiator is different');
		expect(warnMock).toMatchSnapshot();
		warnMock.mockClear();

		expect(excludeDuplicateFiles([
			{
				js: ['first.js', 'second.js'],
				matches: ['https://virgilio.it/*'],
				run_at: 'document_end',
			},
			{
				js: ['first.js', 'second.js'],
				run_at: 'document_start',
				matches: ['https://example.com/*'],
			},
		])).toMatchSnapshot('it should warn when a differentiator is different');
		expect(warnMock).toMatchSnapshot();
		warnMock.mockClear();

		expect(excludeDuplicateFiles([
			{
				css: ['first.css'],
				matches: ['https://virgilio.it/*'],
				all_frames: true,
			},
			{
				css: ['first.css', 'second.css'],
				matches: ['https://example.com/*'],
			},
		])).toMatchSnapshot('it should warn when a differentiator is different');
		expect(warnMock).toMatchSnapshot();
		warnMock.mockClear();
	});
});
