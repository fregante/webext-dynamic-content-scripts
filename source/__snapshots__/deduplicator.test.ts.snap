// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`excludeDuplicateFiles > safe > it should drop the whole block if if empty 1`] = `
[
  {
    "css": [
      "first.css",
    ],
    "js": [
      "alpha.js",
    ],
  },
]
`;

exports[`excludeDuplicateFiles > safe > it should remove duplicate scripts 1`] = `
[
  {
    "css": [],
    "js": [
      "first.js",
    ],
  },
  {
    "css": [],
    "js": [
      "second.js",
    ],
  },
]
`;

exports[`excludeDuplicateFiles > safe > it should remove duplicate scripts and stylesheets 1`] = `
[
  {
    "css": [
      "first.css",
    ],
    "js": [
      "alpha.js",
    ],
  },
  {
    "css": [
      "second.css",
    ],
    "js": [],
  },
]
`;

exports[`excludeDuplicateFiles > safe > it should remove duplicate stylesheets 1`] = `
[
  {
    "css": [
      "first.css",
    ],
    "js": [],
  },
  {
    "css": [
      "second.css",
    ],
    "js": [],
  },
]
`;

exports[`excludeDuplicateFiles > warning > it should not warn when a differentiator is the same 1`] = `
[
  {
    "css": [],
    "exclude_matches": [
      "https://*/admin/*",
    ],
    "js": [
      "first.js",
    ],
  },
]
`;

exports[`excludeDuplicateFiles > warning > it should warn when a differentiator is different 1`] = `
[
  {
    "css": [
      "first.css",
    ],
    "js": [],
  },
  {
    "css": [
      "second.css",
    ],
    "js": [],
    "run_at": "document_start",
  },
]
`;

exports[`excludeDuplicateFiles > warning > it should warn when a differentiator is different 2`] = `
[
  {
    "css": [],
    "js": [
      "first.js",
      "second.js",
    ],
    "run_at": "document_end",
  },
]
`;

exports[`excludeDuplicateFiles > warning > it should warn when a differentiator is different 3`] = `
[
  {
    "all_frames": true,
    "css": [
      "first.css",
    ],
    "js": [],
  },
  {
    "css": [
      "second.css",
    ],
    "js": [],
  },
]
`;

exports[`excludeDuplicateFiles > warning 1`] = `
[MockFunction warn] {
  "calls": [
    [
      "Duplicate file in the manifest content_scripts: first.css 
More info: https://github.com/fregante/webext-dynamic-content-scripts/issues/62",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`excludeDuplicateFiles > warning 2`] = `
[MockFunction warn] {
  "calls": [
    [
      "Duplicate file in the manifest content_scripts: first.js 
More info: https://github.com/fregante/webext-dynamic-content-scripts/issues/62",
    ],
    [
      "Duplicate file in the manifest content_scripts: second.js 
More info: https://github.com/fregante/webext-dynamic-content-scripts/issues/62",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`excludeDuplicateFiles > warning 3`] = `
[MockFunction warn] {
  "calls": [
    [
      "Duplicate file in the manifest content_scripts: first.css 
More info: https://github.com/fregante/webext-dynamic-content-scripts/issues/62",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
